state/state.go

```diff
@@ -1,29 +1,68 @@
 // Copyright © 2019 Arrikto Inc.  All Rights Reserved.
 
-package main
+package state
 
 import (
-	"github.com/gorilla/sessions"
-	"github.com/pkg/errors"
+	"math/rand"
 	"net/http"
 	"net/http/httptest"
 	"time"
+
+	"github.com/gorilla/sessions"
+	"github.com/pkg/errors"
 )
 
 const oidcLoginSessionCookie = "non_existent_cookie"
 
+var nonceChars = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
+
 type state struct {
-	origURL string
+	OrigURL string
 }
 
-func newState(origURL string) *state {
+type Config struct {
+	SchemeDefault string
+	SchemeHeader  string
+	SchemeAndHost bool
+}
+
+type StateFunc func(*http.Request) *state
+
+func NewStateFunc(config *Config) StateFunc {
+	if config.SchemeAndHost {
+		return newSchemeAndHost(config)
+	}
+	return relativeURL
+}
+
+func relativeURL(r *http.Request) *state {
 	return &state{
-		origURL: origURL,
+		OrigURL: r.URL.String(),
+	}
+}
+
+func newSchemeAndHost(config *Config) StateFunc {
+	if config.SchemeHeader == "" {
```

Comment:
Do we need this case at all? If it's empty you can still do r.Header.Get() below on line 54 and it'll return an empty string and you'll use the default scheme anyway. Doesn't seem like we need to special case this case here.

Reply:
We don't really need this case. I was doing this because the default for SCHEME_HEADER was "" and it seems silly to try to read the header in this case. Anyways, it's not much of an optimization to skip checking for the header, so I'll get rid of the closures and have a single function.